README.md
# HiDPI Daemon

This program is for managing HiDPI and LoDPI monitors on X. This program is installed by default in Pop!\_OS and Ubuntu (if installed by System76 and can be added with [this article](https://support.system76.com/articles/system76-software)). 

## Building

### Installing dependencies

`sudo apt install debhelper dh-python python3-all pyflakes3 python3-gi python3-pydbus python3-xlib gir1.2-notify-0.7`

### Build without signing

`sudo dpkg-buildpackage -us uc`

## Removing

This program can be removed with this command:

`sudo apt remove hidpi-daemon`

## Making changes

1. Checkout new branch
2. Push new branch
3. Bump the version (`./bump-version.py`)
4. Make changes
5. Make pull request
6. Get PR approved and merged
7. Make a release from master branch (`./make-release.py`)

## License

This software is made available under the terms of the GNU General Public
License; either version 2 of the License, or (at your option) any later
version. See [LICENSE](LICENSE) for details.



debian/changelog
hidpi-daemon (18.04.7~~alpha) bionic; urgency=low

  * Daily WIP for 18.04.7

 -- Tim Crawford <tcrawford@system76.com>  Fri, 02 Sep 2022 10:26:20 -0600

hidpi-daemon (18.04.6) bionic; urgency=low

  * Update hidpidaemon2.py file to set gi require version
  * Add newer models with HiDPI displays
  * Update Maintainer

 -- Jeremy Soller <jeremy@system76.com>  Tue, 14 Jul 2020 08:49:58 -0600

hidpi-daemon (18.04.5) bionic; urgency=low

  * Fix compilation on eoan

 -- Jeremy Soller <jeremy@system76.com>  Tue, 15 Oct 2019 11:24:55 -0600

hidpi-daemon (18.04.4) bionic; urgency=low

  * Fix layout regression on intel systems
  * Don't override user scaling on session login (unless hidpi is present).
  * Notification now shows correct mode on session login on Intel.
  * Handle laptop lids on more hardware.

 -- David Jordan <david2@system76.com>  Wed, 01 Aug 2018 15:55:45 -0600

hidpi-daemon (18.04.3) bionic; urgency=low

  * Add support for PRIME displays to support laptops with hybrid graphics.
  * Remove workarounds for bugs in the NVIDIA 390 driver
  * Remove gschema override to disable on NVIDIA graphics by default.

 -- David Jordan <david2@system76.com>  Thu, 07 Jun 2018 16:06:53 -0600

hidpi-daemon (18.04.2) bionic; urgency=low

  * Disable notifications when daemon is disabled.
  * hidpi-notification now makes sure dbus interface actually exists.

 -- David Jordan <david2@system76.com>  Thu, 10 May 2018 13:21:21 -0600

hidpi-daemon (18.04.1) bionic; urgency=low

  * Bump for rebuild

 -- Jeremy Soller <jeremy@system76.com>  Tue, 08 May 2018 16:30:42 -0600

hidpi-daemon (18.04.0) bionic; urgency=low

  * Disable on NVIDIA graphics by default using gschema override

 -- David Jordan <david2@system76.com>  Thu, 26 Apr 2018 14:35:00 -0600

hidpi-daemon (17.10.3) artful; urgency=low

  * Workaround: Mutter disables external lowdpi on NVIDIA when in hidpi mode.

 -- David Jordan <david2@system76.com>  Wed, 18 Apr 2018 12:23:32 -0600

hidpi-daemon (17.10.2) artful; urgency=low

  * Projector toggle can now be set to use internal/external displays only
  * Disabled displays remain disabled when set in settings

 -- David Jordan <david2@system76.com>  Thu, 12 Apr 2018 18:58:34 -0600

hidpi-daemon (17.10.1) artful; urgency=low

  * Use deprecated method to disable
  * Never set display modes when disabled via gsetting

 -- David Jordan <david2@system76.com>  Mon, 09 Apr 2018 12:00:22 -0600

hidpi-daemon (17.10.0) artful; urgency=low

  * Initial release

 -- David Jordan <david2@system76.com>  Wed, 04 Apr 2018 16:24:45 -0600


make-release.py
#!/usr/bin/python3

# hidpi-daemon: Universal driver for System76 computers
# Copyright (C) 2005-2016 System76, Inc.
#
# This file is part of `hidpi-daemon`.
#
# `hidpi-daemon` is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# `hidpi-daemon` is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with `hidpi-daemon`; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

"""
Make a stable hidpi-daemon release.
"""

import sys
import os
from os import path
import re
import time
from subprocess import check_call, check_output, call

from hidpidaemon import __version__
from hidpidaemon.tests.helpers import TempDir


DISTROS = ('trusty', 'xenial', 'yakkety', 'zesty', 'artful', 'bionic')
ALPHA = '~~alpha'

TREE = path.dirname(path.abspath(__file__))
assert TREE == sys.path[0]
assert os.getcwd() == TREE

CHANGELOG = path.join(TREE, 'debian', 'changelog')
SETUP = path.join(TREE, 'setup.py')
INIT = path.join(TREE, 'hidpidaemon', '__init__.py')
DSC_NAME =     'hidpi-daemon_{}.dsc'.format(__version__)

assert path.isfile(CHANGELOG)
assert path.isfile(SETUP)
assert path.isfile(INIT)


def confirm():
    while True:
        response = input('  Okay? yes/NO: ').lower()
        if response == 'yes':
            return True
        if response == 'no':
            return False
        print("Please enter 'yes' or 'no'")


def check_for_uncommitted_changes():
    if check_output(['git', 'diff']).decode() != '':
        sys.exit('ERROR: unstaged changes!')
    if check_output(['git', 'diff', '--cached']).decode() != '':
        sys.exit('ERROR: uncommited changes!')


def iter_input_lines(fp):
    yield fp.readline()

    line = fp.readline()
    if line != '\n':
        raise ValueError('bad empty line[1]:\n{!r}'.format(line))
    yield line

    line = fp.readline()
    if not line.startswith('  * Daily WIP for '):
        raise ValueError('bad first item line[2]:\n{!r}'.format(line))

    line = fp.readline()
    if line[:4] != '  * ':
        raise ValueError('bad second item line[3]:\n{!r}'.format(line))
    yield line

    i = 4
    while True:
        line = fp.readline()
        if line[:4] not in ('  * ', '    ', '\n'):
            raise ValueError('bad item line[{}]:\n{!r}'.format(i, line))
        yield line
        i += 1
        if line == '\n':
            break

    line = fp.readline()
    if line[:4] != ' -- ':
        raise ValueError('bad author line[{}]:\n{!r}'.format(i, line))
    yield line


def parse_version_line(line):
    if ALPHA not in line:
        raise ValueError('Missing {!r} in version:\n{!r}'.format(ALPHA, line))
    m = re.match(
        '^hidpi-daemon \(([\.0-9]+)' + ALPHA + '\) ([a-z]+); urgency=low$', line
    )
    if m is None:
        raise ValueError('bad version line[0]:\n{!r}'.format(line))
    ver = m.group(1)
    if ver != __version__:
        raise ValueError(
            'changelog != __version: {!r} != {!r}'.format(ver, __version__)
        )
    distro = m.group(2)
    if distro not in DISTROS:
        raise ValueError('bad distro {!r} not in {!r}'.format(distro, DISTROS))
    return (ver, distro)


def build_version_line(line):
    parse_version_line(line)
    return line.replace(ALPHA, '')


def build_author_line():
    user = check_output(['git', 'config', '--get', 'user.name']).decode().strip()
    email = check_output(['git', 'config', '--get', 'user.email']).decode().strip()
    author = ' '.join((user, '<' + email + '>'))
    ts = time.strftime('%a, %d %b %Y %H:%M:%S %z', time.localtime())
    return ' -- {}  {}\n'.format(author, ts)


def iter_output_lines(input_lines):
    yield build_version_line(input_lines[0])
    yield from input_lines[1:-1]
    yield build_author_line()


# Make sure there are no uncommited changes in the tree:
check_for_uncommitted_changes()

# Read lines from current debian/changelog file:
with open(CHANGELOG, 'r') as fp:
    input_lines = list(iter_input_lines(fp))
    remaining_lines = fp.readlines()

# Parse and validate, then build lines for new changelog files:
(version, distro) = parse_version_line(input_lines[0])
assert version == __version__
output_lines = list(iter_output_lines(input_lines))
assert len(output_lines) == len(input_lines)
assert output_lines[1:-1] == input_lines[1:-1]

# Again, make sure there are no uncommited changes in the tree:
check_for_uncommitted_changes()

# Write the new debian/changelog file:
with open(CHANGELOG, 'w') as fp:
    fp.writelines(output_lines + remaining_lines)

# Make sure the unit tests pass in-tree:
check_call([SETUP, 'test'])

# Make sure package builds okay locally using pbuilder-dist:
check_call(['pbuilder-dist', distro, 'update'])
tmp = TempDir()
os.mkdir(tmp.join('result'))
check_call(['dpkg-source', '-b', TREE], cwd=tmp.join('result'))
check_call(['pbuilder-dist', distro, 'build', tmp.join('result', DSC_NAME)])
del tmp

def abort(msg=None):
    if msg is not None:
        print('\nERROR: ' + msg)
    print('')
    print('Release not made, reverting changes...')
    check_call(['git', 'checkout', '--', CHANGELOG, INIT])
    print('Goodbye.')
    status = (0 if msg is None else 2)
    sys.exit(status)

# Confirm before we make the commit:
print('-' * 80)
call(['git', 'diff'])
print('-' * 80)
print('Source tree is {!r}'.format(TREE))
print('Will release {!r} for {!r}'.format(version, distro))
if not confirm():
    abort()

# Commit and tag:
check_call(['git', 'commit', CHANGELOG, '-m', 'Release {}'.format(version)])
check_call(['git', 'push'])
check_call(['git', 'tag', version])
check_call(['git', 'push', 'origin', 'tag', version])

# We're done:
print('-' * 80)
print('Released {!r} for {!r}'.format(version, distro))


hidpidaemon/xlib.py
# hidpi-daemon: HiDPI daemon to manage HiDPI and LoDPI monitors on X
# Copyright (C) 2017-2018 System76, Inc.
#
# This file is part of `hidpi-daemon`.
#
# `hidpi-daemon` is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# `hidpi-daemon` is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with `hidpi-daemon`; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

"""
Patches for Xlib bindings.  The version included in debian-based distros (0.14) was several years out of date since python-xlib moved from sourceforge to github.  Not needed for releases that include 0.20.
"""

from Xlib import X
from Xlib.ext import randr
from Xlib.protocol import rq

extname = 'RANDR'

class _GetOutputInfo(rq.ReplyRequest):
    _request = rq.Struct(
        rq.Card8('opcode'),
        rq.Opcode(9),
        rq.RequestLength(),
        rq.Card32('output'),
        rq.Card32('config_timestamp'),
        )
    _reply = rq.Struct(
        rq.ReplyCode(),
        rq.Card8('status'),
        rq.Card16('sequence_number'),
        rq.ReplyLength(),
        rq.Card32('timestamp'),
        rq.Card32('crtc'),
        rq.Card32('mm_width'),
        rq.Card32('mm_height'),
        rq.Card8('connection'),
        rq.Card8('subpixel_order'),
        rq.LengthOf('crtcs', 2),
        rq.LengthOf('modes', 2),
        rq.Card16('num_preferred'),
        rq.LengthOf('clones', 2),
        rq.LengthOf('name', 2),
        rq.List('crtcs', rq.Card32Obj),
        rq.List('modes', rq.Card32Obj),
        rq.List('clones', rq.Card32Obj),
        rq.String8('name'),
)

def _get_output_info(d, output, config_timestamp):
    return _GetOutputInfo(
        display=d.display,
        opcode=d.display.get_extension_major(extname),
        output=output,
        config_timestamp=config_timestamp,
)


class _GetCrtcInfo(rq.ReplyRequest):
    _request = rq.Struct(
        rq.Card8('opcode'),
        rq.Opcode(20),
        rq.RequestLength(),
        rq.Card32('crtc'),
        rq.Card32('config_timestamp'),
        )
    _reply = rq.Struct(
        rq.ReplyCode(),
        rq.Card8('status'),
        rq.Card16('sequence_number'),
        rq.ReplyLength(),
        rq.Card32('timestamp'),
        rq.Int16('x'),
        rq.Int16('y'),
        rq.Card16('width'),
        rq.Card16('height'),
        rq.Card32('mode'),
        rq.Card16('rotation'),
        rq.Card16('possible_rotations'),
        rq.LengthOf('outputs', 2),
        rq.LengthOf('possible_outputs', 2),
        rq.List('outputs', rq.Card32Obj),
        rq.List('possible_outputs', rq.Card32Obj),
        )

def _get_crtc_info(d, crtc, config_timestamp):
    return _GetCrtcInfo (
        display=d.display,
        opcode=d.display.get_extension_major(extname),
        crtc=crtc,
        config_timestamp=config_timestamp,
)


class _CreateMode(rq.ReplyRequest):
    _request = rq.Struct(
        rq.Card8('opcode'),
        rq.Opcode(16),
        rq.RequestLength(),
        rq.Window('window'),
        rq.Object('mode', randr.RandR_ModeInfo),
        rq.String8('name'),
        )
    _reply = rq.Struct(
        rq.ReplyCode(),
        rq.Pad(1),
        rq.Card16('sequence_number'),
        rq.ReplyLength(),
        rq.Card32('mode'),
        rq.Pad(20),
        )

def _create_mode(w, mode, name):
    return _CreateMode (
        display=w.display,
        opcode=w.display.get_extension_major(extname),
        window=w,
        mode=mode,
        name=name,
)


class _AddOutputMode(rq.Request):
    _request = rq.Struct(
        rq.Card8('opcode'),
        rq.Opcode(18),
        rq.RequestLength(),
        rq.Card32('output'),
        rq.Card32('mode'),
        )

def _add_output_mode(d, output, mode):
    return _AddOutputMode(
        display=d.display,
        opcode=d.display.get_extension_major(extname),
        output=output,
        mode=mode,
)


class _SetCrtcConfig(rq.ReplyRequest):
    _request = rq.Struct(
        rq.Card8('opcode'),
        rq.Opcode(21),
        rq.RequestLength(),
        rq.Card32('crtc'),
        rq.Card32('timestamp'),
        rq.Card32('config_timestamp'),
        rq.Int16('x'),
        rq.Int16('y'),
        rq.Card32('mode'),
        rq.Card16('rotation'),
        rq.Pad(2),
        rq.List('outputs', rq.Card32Obj),
        )
    _reply = rq.Struct(
        rq.ReplyCode(),
        rq.Card8('status'),
        rq.Card16('sequence_number'),
        rq.ReplyLength(),
        rq.Card32('new_timestamp'),
        rq.Pad(20),
        )

def _set_crtc_config(d, crtc, config_timestamp, x, y, mode, rotation, outputs, timestamp=X.CurrentTime):
    return _SetCrtcConfig (
        display=d.display,
        opcode=d.display.get_extension_major(extname),
        crtc=crtc,
        config_timestamp=config_timestamp,
        x=x,
        y=y,
        mode=mode,
        rotation=rotation,
        outputs=outputs,
        timestamp=timestamp,
)

class _GetOutputProperty(rq.ReplyRequest):
    _request = rq.Struct(
        rq.Card8('opcode'),
        rq.Opcode(15),
        rq.RequestLength(),
        rq.Card32('output'),
        rq.Card32('property'),
        rq.Card32('type'),
        rq.Card32('long_offset'),
        rq.Card32('long_length'),
        rq.Bool('delete'),
        rq.Bool('pending'),
        rq.Pad(2),
        )
    _reply = rq.Struct(
        rq.ReplyCode(),
        rq.Format('value', 1),
        rq.Card16('sequence_number'),
        rq.ReplyLength(),
        rq.Card32('property_type'),
        rq.Card32('bytes_after'),
        rq.LengthOf('value', 4),
        rq.Pad(12),
        rq.List('value', rq.Card8Obj),
        )

def _get_output_property(d, output, property, type, long_offset, long_length, delete=False, pending=False):
    print('get output property override')
    return _GetOutputProperty(
        display=d.display,
        opcode=d.display.get_extension_major(extname),
        output=output,
        property=property,
        type=type,
        long_offset=long_offset,
        long_length=long_length,
        delete=delete,
        pending=pending,
)


hidpidaemon/__init__.py
# hidpi-daemon: HiDPI daemon to manage HiDPI and LoDPI monitors on X
# Copyright (C) 2017-2018 System76, Inc.
#
# This file is part of `hidpi-daemon`.
#
# `hidpi-daemon` is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# `hidpi-daemon` is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with `hidpi-daemon`; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

"""
HiDPI daemon to manage HiDPI and LoDPI monitors on X.
"""

from os import path
import logging


__version__ = '18.04.7'

datadir = path.join(path.dirname(path.abspath(__file__)), 'data')
log = logging.getLogger(__name__)


# Unfortunately, we need to accomidate some typos and goofs in sys_vendor:
VALID_SYS_VENDOR = (
    'System76',             # Current standard
    'System76, Inc.',       # Previous standard
    'System76, Inc',
    'System76, Inc .',
    'Notebook',
)


def get_datafile(name):
    return path.join(datadir, name)


def read_dmi_id(key, sysdir='/sys'):
    if key not in ('sys_vendor', 'product_version'):
        raise ValueError('bad dmi/id key: {!r}'.format(key))
    filename = path.join(sysdir, 'class', 'dmi', 'id', key)
    try:
        with open(filename, 'r') as fp:
            return fp.read(256).strip()
    except (FileNotFoundError, UnicodeDecodeError):
        pass


def get_sys_vendor(sysdir='/sys'):
    sys_vendor = read_dmi_id('sys_vendor', sysdir)
    if sys_vendor in VALID_SYS_VENDOR:
        return sys_vendor
    log.warning('invalid sys_vendor: %r', sys_vendor)


def get_product_version(sysdir='/sys'):
    if get_sys_vendor(sysdir) is not None:
        return read_dmi_id('product_version', sysdir)


hidpidaemon/dbusutil.py
# hidpi-daemon: HiDPI daemon to manage HiDPI and LoDPI monitors on X
# Copyright (C) 2017-2018 System76, Inc.
#
# This file is part of `hidpi-daemon`.
#
# `hidpi-daemon` is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# `hidpi-daemon` is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with `hidpi-daemon`; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

"""
DBus utility functions for the hidpi daemon.  On Gnome we need to talk to Mutter over its dbus api to set scaling on nvidia systems.
"""

from gi.repository import Gio, GLib


def dbus_helper(destination = 'org.gnome.Mutter.DisplayConfig',
                path        = '/org/gnome/Mutter/DisplayConfig',
                interface   = 'org.gnome.Mutter.DisplayConfig',
                method      = None,
                args        = None,
                answer_fmt  = None,
                proxy_prpty = Gio.DBusCallFlags.NONE,
                timeout     = -1,
                cancellable = None 
                ):
    
    bus = Gio.bus_get_sync(Gio.BusType.SESSION, None)
    reply = bus.call_sync(destination, path, interface,
                      method, args, answer_fmt,
                      proxy_prpty, timeout, cancellable)
    return reply

def unpack_current_state(current_state):
    configuration_serial = current_state[0]
    monitors = current_state[1]
    log_displays = current_state[2]
    logical_displays = []
    for log_display in log_displays:
        logical_display = { 'x_position' : log_display[0],
                            'y_position' : log_display[1],
                            'scale'      : log_display[2],
                            'transform'  : log_display[3],
                            'primary'    : log_display[4]
                          }
        #logical_displays.append(logical_display)
        log_monitors = log_display[5]
        processed_monitors = []
        for monitor in log_monitors:
            processed_monitor = {}
            processed_monitor['connector'] = monitor[0]
            processed_monitor['vendor']    = monitor[1]
            processed_monitor['product']   = monitor[2]
            processed_monitor['serial']    = monitor[3]
            for mon in monitors:
                if mon[0][0] == monitor[0]:
                    processed_monitor['modes'] = mon[1]
            processed_monitors.append(processed_monitor)
        logical_display['monitors'] = processed_monitors
        logical_displays.append(logical_display)
    return configuration_serial, logical_displays

def get_current_state():
    current_state = dbus_helper(method='GetCurrentState', 
                                answer_fmt  = GLib.VariantType.new ('(ua((ssss)a(siiddada{sv})a{sv})a(iiduba(ssss)a{sv})a{sv})')
    )
    
    return unpack_current_state(current_state)
    
def apply_monitors_configuration(configuration_serial, displays, scale):
    displays_arg = []
    
    for display in displays:
        #generate monitors argument for each display
        monitors_arg = []
        display_scale = scale
        for monitor in display['monitors']:
            monitors_arg.append(( 
                        monitor['connector'], 
                        monitor['modes'][0][0], 
                        {
                            'underscanning': GLib.Variant('b', False)
                        }
                    ))
            if scale not in monitor['modes'][0][5]:
                display_scale = 1.0
        display_arg = (
                    display['x_position'], 
                    display['y_position'], 
                    display_scale, 
                    display['transform'], 
                    display['primary'], 
                    monitors_arg, 
                )
        displays_arg.append(display_arg)
    args = GLib.Variant('(uua(iiduba(ssa{sv}))a{sv})', ( configuration_serial, 1,
            displays_arg,
                {
                },
            )
            )
    dbus_helper(method='ApplyMonitorsConfig',
                    args = args)

def get_scale():
    configuration_serial, displays = get_current_state()
    scale = 1.0
    for display in displays:
        if display['scale'] > scale:
            scale = display['scale']
    return scale

def set_scale(scale):
    configuration_serial, displays = get_current_state()
    apply_monitors_configuration(configuration_serial, displays, scale)
    
#set_scale(2.0)


bump-version.py
#!/usr/bin/python3

# hidpi-daemon: Universal driver for System76 computers
# Copyright (C) 2005-2018 System76, Inc.
#
# This file is part of `hidpi-daemon`.
#
# `hidpi-daemon` is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# `hidpi-daemon` is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with `hidpi-daemon`; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

"""
Bump version of hidpi-daemon to start work on next release.
"""

import sys
import os
from os import path
import re
import time
from subprocess import check_call, check_output, call

from hidpidaemon import __version__
from hidpidaemon.tests.helpers import TempDir


DISTROS = ('trusty', 'xenial', 'yakkety', 'zesty', 'artful', 'bionic', 'cosmic', 'disco', 'eoan', 'focal')
ALPHA = '~~alpha'

TREE = path.dirname(path.abspath(__file__))
assert TREE == sys.path[0]
assert os.getcwd() == TREE

CHANGELOG = path.join(TREE, 'debian', 'changelog')
INIT = path.join(TREE, 'hidpidaemon', '__init__.py')
SETUP = path.join(TREE, 'setup.py')

assert path.isfile(CHANGELOG)
assert path.isfile(INIT)
assert path.isfile(SETUP)


def confirm():
    while True:
        response = input('  Okay? yes/NO: ').lower()
        if response == 'yes':
            return True
        if response == 'no':
            return False
        print("Please enter 'yes' or 'no'")


def check_for_uncommitted_changes():
    if check_output(['git', 'diff']).decode() != '':
        sys.exit('ERROR: unstaged changes!')
    if check_output(['git', 'diff', '--cached']).decode() != '':
        sys.exit('ERROR: uncommited changes!')


def parse_version_line(line):
    if ALPHA in line:
        raise ValueError('{!r} in current version:\n{!r}'.format(ALPHA, line))
    m = re.match(
        '^hidpi-daemon \(([\.0-9]+)\) ([a-z]+); urgency=low$', line
    )
    if m is None:
        raise ValueError('bad version line[0]:\n{!r}'.format(line))
    ver = m.group(1)
    if ver != __version__:
        raise ValueError(
            'changelog != __version: {!r} != {!r}'.format(ver, __version__)
        )
    distro = m.group(2)
    if distro not in DISTROS:
        raise ValueError('bad distro {!r} not in {!r}'.format(distro, DISTROS))
    return (ver, distro)


def bump_version(current):
    (year, month, rev) = current.split('.')
    assert str(int(year)) == year and int(year) >= 14
    assert month in ('04', '10')
    assert str(int(rev)) == rev and int(rev) >= 0
    return '{}.{}.{}'.format(year, month, int(rev) + 1)


def build_author_line():
    user = check_output(['git', 'config', '--get', 'user.name']).decode().strip()
    email = check_output(['git', 'config', '--get', 'user.email']).decode().strip()
    author = ' '.join((user, '<' + email + '>'))
    ts = time.strftime('%a, %d %b %Y %H:%M:%S %z', time.localtime())
    return ' -- {}  {}\n'.format(author, ts)


def iter_new_changelog_lines(new, newdeb, distro):
    yield 'hidpi-daemon ({}) {}; urgency=low\n'.format(newdeb, distro)
    yield '\n'
    yield '  * Daily WIP for {}\n'.format(new)
    yield '\n'
    yield build_author_line()
    yield '\n'


def iter_new_init_lines(new, init_lines):
    found_version = False
    for line in init_lines:
        if line.startswith('__version__'):
            assert found_version is False
            found_version = True
            yield '__version__ = {!r}\n'.format(new)
        else:
            yield line


# Make sure there are no uncommited changes in the tree:
check_for_uncommitted_changes()

# Read existing changelog and __init__.py lines:
with open(CHANGELOG, 'r') as fp:
    changelog_lines = fp.readlines()
with open(INIT, 'r') as fp:
    init_lines = fp.readlines()

# Parse out current version and distro, bump version:
(current, distro) = parse_version_line(changelog_lines[0])
assert current == __version__
new = bump_version(current)
assert new != __version__
newdeb = new + ALPHA

# Build new changelog and __init__.py lines:
new_changelog_lines = list(iter_new_changelog_lines(new, newdeb, distro))
new_init_lines = list(iter_new_init_lines(new, init_lines))
assert len(new_changelog_lines) == 6
assert len(new_init_lines) == len(init_lines)

# Again, make sure there are no uncommited changes in the tree:
check_for_uncommitted_changes()

# Write new changelog and __init__.py lines:
with open(CHANGELOG, 'w') as fp:
    fp.writelines(new_changelog_lines + changelog_lines)
with open(INIT, 'w') as fp:
    fp.writelines(new_init_lines)

# Confirm before we make the commit:
print('-' * 80)
call(['git', 'diff'])
print('-' * 80)
print('Source tree is {!r}'.format(TREE))
print(
    'Will bump {!r} version from {!r} to {!r}'.format(distro, current, newdeb)
)
if not confirm():
    print('')
    print('Version bump not committed, reverting changes...')
    check_call(['git', 'checkout', '--', CHANGELOG, INIT])
    print('Goodbye.')
    sys.exit(0)

# Make the commit:
check_call(['git', 'commit', CHANGELOG, INIT, '-m', 'Bump version to {}'.format(newdeb)])
check_call(['git', 'push'])
print('-' * 80)
print('{!r} is now at version {!r}'.format(distro, newdeb))
